### PR TITLE
Encrypted parameters

### PR DESCRIPTION
**Pull request by mfleisch**
### Implementation of parameter encryption and sensitive parameter handling
* Parameters can be saved AES encrypted using a user/config specific encryption key
* Parameters can be sensitive. Sensitive Parameters will not be written into project configuration.xml (e.g. Passwords)
* Changed the location of the mzmine config.xml file to the user home directory

#### Sensitive Parameters:
* Error email config is declared as sensitive
* GNPS login credential are declared as sensitive
* The PasswordParameter is encrypted and sensitive

### CODE REVIEW

File: src/main/java/net/sf/mzmine/util/StringCrypter.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
+package net.sf.mzmine.util;
```
Comment (tomas-pluskal) [inline src/main/java/net/sf/mzmine/util/StringCrypter.java]: Could you please copy the license header to all new .java files?

File: src/main/java/net/sf/mzmine/parameters/parametertypes/PasswordParameter.java [hunk 2]
```diff
@@ -83,7 +87,7 @@ public void setValue(String newValue) {
   public boolean checkValue(Collection<String> errorMessages) {
     if (!valueRequired)
       return true;
-    if ((value == null) || (value.trim().length() == 0)) {
+    if ((value == null) || (value.isEmpty())) { //white spaces can be valid in pw. So no trim
```
Comment (tomas-pluskal) [inline src/main/java/net/sf/mzmine/parameters/parametertypes/PasswordParameter.java]: It's true that pw can contain white space, but don't you think a password that consists only of whitespace should be considered invalid? :)

File: src/main/java/net/sf/mzmine/util/StringCrypter.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+
+import de.unijena.bioinf.babelms.utils.Base64;
```
Comment (tomas-pluskal) [inline src/main/java/net/sf/mzmine/util/StringCrypter.java]: Could you switch this to use java.util.Base64 class instead the de.unijena class?

File: src/main/java/net/sf/mzmine/parameters/impl/SimpleParameterSet.java [hunk 4]
```diff
@@ -59,6 +60,14 @@ public SimpleParameterSet(Parameter<?> parameters[]) {
     return parameters;
   }
 
+  public void setSkipSensitiveParameters(boolean skipSensitiveParameters) {
```
Comment (tomas-pluskal) [inline src/main/java/net/sf/mzmine/parameters/impl/SimpleParameterSet.java]: Is it useful to have the skipSensitiveParameters as a field of this class (i.e., a state that is remembered)? I would simply add it as a parameter of saveValuesToXML()

**Comment by tomas-pluskal** [comment]
Thanks Markus, it looks good. I added a few minor comments.

File: src/main/java/net/sf/mzmine/parameters/impl/SimpleParameterSet.java (hunk 4 above)
Comment (mfleisch) [reply src/main/java/net/sf/mzmine/parameters/impl/SimpleParameterSet.java]: I agree that Parameter and ParameterSet objects should be kept lightweight.

But if we just add it as a parameter of saveValuesToXML(), it is not so easy to inherit this information to Parameters that contain ParameterSets (e.g. ParameterSetParameter).

To do so we would have to add the skipSensitive boolean to the saveValueToXML() method of the Parameters.

The disadvantage of this solution is that we would have to check the skipSensitive value twice. One time in the ParameterSet (to not write an empty xml block of skipped parameters) and another time within the saveValueToXML() method of sensitive parameters. Which I found less consistent.

In general, I am fine with both solutions. Let me know which you prefer.

File: src/main/java/net/sf/mzmine/util/StringCrypter.java (hunk 1 above)
Comment (mfleisch) [reply src/main/java/net/sf/mzmine/util/StringCrypter.java]: Done :-)

File: src/main/java/net/sf/mzmine/parameters/parametertypes/PasswordParameter.java (hunk 2 above)
Comment (mfleisch) [reply src/main/java/net/sf/mzmine/parameters/parametertypes/PasswordParameter.java]: I thought that it is a bit strange to check the password requirements of third party applications.
However, all whitespace passwords are a strange edge case. I reverted the change.

File: src/main/java/net/sf/mzmine/util/StringCrypter.java (hunk 3 above)
Comment (mfleisch) [reply src/main/java/net/sf/mzmine/util/StringCrypter.java]: Of course, this was by accident :-)

**Comment by tomas-pluskal** [comment]
Looks good, thanks!

**Pull request merged**